x-pack/plugins/transform/public/app/hooks/use_index_data.ts

```diff
@@ -92,20 +97,39 @@ export const useIndexData = (
     } catch (e) {
       setErrorMessage(getErrorMessage(e));
       setStatus(INDEX_STATUS.ERROR);
+      return;
     }
   };
 
+  const fetchColumnChartsData = async function () {
+    const columnChartsData = await fetchChartsData(
+      indexPattern.title,
+      api,
+      isDefaultQuery(query) ? matchAllQuery : query,
+      columns.filter((cT) => dataGrid.visibleColumns.includes(cT.id))
+    );
+
+    setColumnCharts(columnChartsData);
+  };
+
   useEffect(() => {
-    getIndexData();
+    fetchDataGridData();
     // custom comparison
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [indexPattern.title, JSON.stringify([query, pagination, sortingColumns])]);
 
+  useEffect(() => {
+    if (chartsVisible) {
+      fetchColumnChartsData();
+    }
+    // custom comparison
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [chartsVisible, indexPattern.title, JSON.stringify([query, dataGrid.visibleColumns])]);
```

Comment:
please avoid using `JSON.stringify`

Reply:
This approach is taken from the same file for another hook, so I'd like to keep it as is for this PR and be consistent within the file. If we want to get rid of the `stringify` we should do it for the whole file, but that may affect more related code. We originally started using `JSON.stringify()` for comparing simple objects where we are not in full control how they look (e.g. ES queries) and deep value checks don't always make sense. Also note that the React team doesn't necessarily suggest to avoid its use completely, there are just some caveats to keep in mind, see `Option 3: (be careful) JSON.stringify` in this comment: https://github.com/facebook/react/issues/14476#issuecomment-471199055

Reply:
Thanks for sharing! But I still believe updating references only when the data has been changed is a better approach 🙂 But it can be resolved in a follow-up